be/src/exec/spill/dir_manager.cpp

```diff
@@ -101,9 +102,20 @@ Status DirManager::init(const std::string& spill_dirs) {
 }
 
 StatusOr<Dir*> DirManager::acquire_writable_dir(const AcquireDirOptions& opts) {
-    // @TODO(silverbullet233): refine the strategy for dir selection
-    size_t idx = _idx++ % _dirs.size();
-    return _dirs[idx].get();
+    // for the case of multiple dirs, we randomly select one as the start
+    // and then try one by one until we find the first one that meets the capacity requirements.
+    size_t start_idx = 0;
+    if (_dirs.size() > 1) {
+        std::lock_guard l(_mutex);
+        start_idx = _rand.Next() % _dirs.size();
+    }
+    for (size_t i = 0; i < _dirs.size(); i++) {
+        size_t idx = (start_idx + i) % _dirs.size();
+        if (_dirs[idx]->inc_size(opts.data_size)) {
```

Comment:
At this point, does the chosen dir is really written into already? if have not,  can we adjust the data size of the dir to right state if writing fails later.

Reply:
> At this point, does the chosen dir is really written into already? if have not, can we adjust the data size of the dir to right state if writing fails later.

At this time, the data is not actually written to the directory, but the amount of data that needs to be written can be clearly known. The current approach is to ensure that after the allocation is successful, the writing will be successful too. If the processing fails when writing, it will be more troublesome to implement, and after the introduction of remote block, things will become more complicated.